push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,190 @@
+In UPS, the developer creates a 'Push Application' and 1 or more 'Variants' to represent the App they are developing outside of OpenShift.
+There is duplication of logical constructs and potential confusion between what Mobile Client is and what a Push Application is.
+
+The developer also has to somehow get the Push configuration from UPS and place it into their App for consumption by the SDK. With the mobile-services.json approach being used by the Mobile CLI/UI and the SDKs for configuration, there's some things that can be done to simplify what the developer has to do here.
+
+*As a developer, I want to write custom logic for sending Push messages so that I can dynamically send messages based on varying factors and inputs*
+
+Push messages can be sent by calling the [sender API](https://aerogear.org/docs/specs/aerogear-unifiedpush-rest/index.html#397083935) with the correct key. There are features available in the OpenShift and the Catalog that can simplify the provisoining of a custom service, and obtaining secret credentials (like a key) from a provisioned service.
+
+## OAuth provider support 
+
+To address the first & second stories, UPS will be modified to support different auth providers.
+The 2 providers will be Keycloak (to maintain backwards compatibility for existing users) and a generic OAuth Provider.
+The OAuth flow will follow the same pattern & flow as [Prometheus](https://github.com/aerogearcatalog/metrics-apb/blob/9b4cb90988f5f3e7a28a84050cf274355cd36498/roles/provision-metrics-apb/tasks/provision-prometheus.yml#L36-L52) with the [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy).
+The UPS APB will be modified to include an OAuth ServiceAccount and the OAuth Proxy container.
+UPS will be modified to allow running in 'no auth' mode i.e. the UI & API will be uprotected and give 'admin' access to the UPS instance running in the user's namespace. The OAuth Proxy will sit in front of UPS providing authentication and autorisation based on namespace permissions.
+
+A future change to this could look at adding support for individual 'developer' users.
+This would be required if running UPS as a shared service.
+That would most likely require changes to UPS to call OpenShift to check user permissions based on the token header, and only showing that developers Push Applications.
+
+These changes will mean UPS will no longer have a dependency on Keycloak when running in OpenShift.
+Also, the developer doesn't have to know about or be concerned with where or how the auth provider is setup and configured for UPS.
+
+## OpenShift UI Integration
+
+To address the third story, a number of UI screens and integration points will be added to the OpenShift UI & the UPS APB.
+
+Push Variant credentials will be stored in OpenShift as a Secret.
+The Secret will have:
+
+* a label to identify it as storing Variant credentials
+* an annotation to link to a specific Mobile Client `aerogear.org/mobile-client-id`
```

Comment:
How would the mobile client discover the secrets associated with it? Should this be a label instead or can we also write selectors over annotations?

Reply:
Label makes more sense. Updating